### DISCUSSION TITLE
Multi-column sorting for Tables

### DISCUSSION THREAD

**Discussion by runrc** [cogentcore/core#1076] [Ideas]
To enhance the usability of Tables, it would be useful to allow multi-column sorting via selection of multiple column headers. Clicking on a column header would allow up, down or no-sorting. Subsequent columns could be added to the sorting criteria by clicking. Columns can be removed by clicking the column header to select no-sorting.

**Comment by kkoreilly** [comment]
So the first column you click on is the primary sorting criteria, and then each additional column you click on is used as a secondary sorting criteria to sort within each resultant sorting group of the first column? Also, currently when you click on a column already used for sorting, it switches back and forth between sorting ascending and descending; how do you remove columns from sorting then? Is it a third option that you click through? I can envision clicking through multiple options being somewhat confusing, so maybe we would put the option to stop sorting in a context menu on the header? Regardless, this seems like a reasonable possibility. Thank you for the suggestion!

**Reply by runrc** [reply]
Yes. The sorting would be tri-state, up, down, or no-sorting. The default being no-sorting, as per initialisation.
Imagine a table of 3 columns - Country, Name, Age. You would be able to sort between say Age first, Country second, then Name.
Selecting no-sorting would exclude it from the sorting criteria.
We may need to add a sub-script number to the indicator to indicate the order of sorting of the column.

**Reply by rcoreilly** [reply]
another option to consider for multi column sorting would be the standard extend-select dynamic, with a control / command or shift key along with the click.

but being able to turn off sorting easily with the tri-state dynamic seems like an improvement, and seems intuitive.  if you have that, then it seems reasonable to have that operate independently on each header, instead of an implied mutex.

**Reply by kkoreilly** [reply]
Yep, if we have independent tri-state for each column, I don't think we need any separate extend select thing; you should just be able to sort each column header without mutex. It would be good to have a nice way to indicate the sorting precedence order.

**Reply by kkoreilly** [reply]
I can create a feature request issue from this discussion; this isn't the absolute highest priority for us, but it shouldn't be too hard to implement, and if you have a use case for it @runrc, we can implement it soon (or you can try a PR if you want of course). (See #1079).

**Reply by runrc** [reply]
OK great. We're dealing with hundreds of thousands of rows of data. Thus, sorting via multiple criteria would be very useful.

**Reply by kkoreilly** [reply]
OK. Feel free to ping us later when you need it if it hasn't been implemented yet.

**Reply by rcoreilly** [reply]
@runrc you might want to look into the tensor/table data structure given the scale of your data -- it has a very efficient index-based sorting algorithm and built-in support for multi-column sorting.  It uses the same basic gui to display the table.  Good thing we made the table only use the visible rows and scroll only the data :)

**Reply by rcoreilly** [reply]
also has interactive filtering (which we can improve now that we have yaegi integrated)

**Reply by kkoreilly** [reply]
Just to be clear, it is `tensor/tensorcore.Table`. If you need any help with that API, you can ask; it is not as well documented yet.
